source/lesson-2-ruby-and-the-command-line.html.md.erb

```diff
@@ -0,0 +1,722 @@
+
+Just like in the command line, if you give `irb` a command it doesn't
+understand it will show you an error message. Again, don't panic! Everything
+is fine.
+
+`irb` will do its best to explain why it didn't understand your ruby, but
+sometimes the error messages can be hard for a beginner to understand.
+
+For example:
+
+```
+irb(main):018:0> "Hello IRB" + 7
+TypeError: no implicit conversion of Fixnum into String
+        from (irb):1:in `+'
+        from (irb):1
+        from /usr/bin/irb:11:in `<main>'
+```
+
+Here, `irb` is telling us that you're not allowed to add strings and numbers
+together in ruby.
+
+### Variables
+
+Programming is all about creating abstractions, and in order to create an
+abstraction we must be able to assign names to things. Variables are a way of
+creating a name for a piece of data.
+
+Creating and using variables in ruby is super simple, it's just `name` `=`
+`expression`. Some examples:
+
+```ruby
+name = "Fido"
+age_human_years = 4
+age_dog_years = age_human_years * 7
+```
+
+Variable names in ruby have to start with a letter, and they can't contain
+spaces or "special" characters like `-`, `$`, `@` and `&`.
+
+As a style convention, ruby variables use underscores to separate the bits of
+the name - this is called `snake_case` (as opposed to `camelCase` or `PascalCase` which are used elsewhere).
+
+### String interpolation
+
+We often need to build a string out of other strings and ruby expressions. In
+ruby we can do this with "string interpolation".
+
+Within the string we use the interpolation marker `#{}`. Inside those
+brackets we can put any variables or Ruby code which will be evaluated,
+converted to a string, and output in that spot of the outer string.
+
+```
+irb(main):019:0> "The dog #{name} is #{age_human_years} human years old, which is #{age_dog_years} in dog years."
+=> "The dog Fido is 4 human years old, which is 28 in dog years."
+```
+
+> Note: on macOs you can enter the `#` key with option (`⌥`) + 3
+
+### Calling methods
+
+While programming, we often find ourselves doing the same thing over and over
+again. It would be nice if we could give a particular task a name, and run it
+just by calling its name.
+
+In ruby we do this with *methods*, and there are lots of them already built
+in to the language. One example is `rand()`, which generates a random number:
+
+```
+irb(main):020:0> rand()
+=> 0.11487911496307956
+```
+
+If we want to generate a random number less than a particular number, we can
+provide our maximum to `rand()` as an "argument":
+
+```
+irb(main):021:0> rand(100)
+=> 7
+irb(main):022:0> rand(100)
+=> 43
+```
+
+You can also call some methods on *things* by using a `.` like this:
+
+```
+irb(main):023:0> "make me louder".upcase()
+=> "MAKE ME LOUDER"
+irb(main):024:0> "em esrever".reverse()
+=> "reverse me"
+```
+
+#### Task 3: Decode a secret message
+
+Decode this top secret, encrypted message:
+
+```
+!gniog peeK !llew repus gniod er'uoY
```

Comment:
As your resident Welsh person, I can confirm that `llew` is lion 😉

Reply:
Diolch yn fawr!